File: leetsync/3555-final-array-state-after-k-multiplication-operations-i/final-array-state-after-k-multiplication-operations-i.java
class Solution {
    public int[] getFinalState(int[] nums, int k, int multiplier) {
        int n = nums.length;
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> {
            int valueComparision = Integer.compare(a[0], b[0]);
            if (valueComparision == 0) {
                return Integer.compare(a[1], b[1]);
            }
            return valueComparision;
        });
        for (int i = 0; i < n; i++) {
            pq.add(new int[] { nums[i], i });
        }
        while (k-- > 0) {
            int[] current = pq.poll();
            int num = current[0];
            int index = current[1];
            nums[index] *= multiplier;
            current[0] *= multiplier;
            pq.add(current);
        }
        return nums;
    }
}

File: leetsync/3555-final-array-state-after-k-multiplication-operations-i/final-array-state-after-k-multiplication-operations-i.go
type minHeap []*numsIndex

type numsIndex struct {
	nums  []int
	index int
}

func (h minHeap) Len() int {
	return len(h)
}

func (h minHeap) Swap(i, j int) {
	h[i], h[j] = h[j], h[i]
}

func (h minHeap) Less(i, j int) bool {
	return h[i].nums[h[i].index] < h[j].nums[h[j].index] ||
		(h[i].nums[h[i].index] == h[j].nums[h[j].index] && h[i].index < h[j].index)
}

func (h *minHeap) Push(obj any) {
	*h = append(*h, obj.(*numsIndex))
}

func (h *minHeap) Pop() any {
	old := *h
	length := len(old)
	element := old[length-1]
	*h = old[:length-1]
	return element
}

func getFinalState(nums []int, k int, multiplier int) []int {
	var arrIndexNums = make(minHeap, 0, len(nums))
	for i := 0; i < len(nums); i++ {
		arrIndexNums = append(arrIndexNums, &numsIndex{nums, i})
	}
	heap.Init(&arrIndexNums)

	for i := 0; i < k; i++ {
		arrIndexNums[0].nums[arrIndexNums[0].index] *= multiplier
		heap.Fix(&arrIndexNums, 0)
	}
	return nums
}